source/language/openpulse.rst

```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
 
-   // amp is pulse amplitude at center
-   // center is the mean of pulse
-   // square_width is the width of the square pulse component
-   // sigma is the standard deviation of pulse
+   // amp is waveform amplitude at center
+   // center is the mean of waveform
+   // square_width is the width of the square waveform component
+   // sigma is the standard deviation of waveform
    gaussian_square(length:l, complex[float[32]]:amp, length:center, length:square_width, length:sigma)
 
-   // amp is pulse amplitude at center
-   // center is the mean of pulse
-   // sigma is the standard deviation of pulse
+   // amp is waveform amplitude at center
+   // center is the mean of waveform
+   // sigma is the standard deviation of waveform
    // beta is the Y correction amplitude, see the DRAG paper
    drag(length:l, complex[float[32]]:amp, length:center, length:sigma, float[32]:beta)
 
-   // Define a constant pulse of length l
+   // Define a constant waveform of length l
    constant(l:length)
 
+   // Define a sine wave of a given amplitude, frequncy, phase, and length
+   sine(l: length, complex[float[32]]:amp, float[32]: frequency, angle: phase)
+
+We can manipulate the ``waveform`` types using the following signal processing functions to produce
+new waveforms
+
+- ``mix(wf1: waveform, wf2: waveform)`` -> waveform - Mix two input waveforms to produce a new waveform.
+  This is equivalent to the product signal :math:`wf(t_i) = wf_1(t_i) \times wf_2(t_i)`
+- ``sum(wf1: waveform, wf2: waveform)`` -> waveform - Sum two input waveforms to produce a new waveform.
+- ``piecewise(wf0: waveform, wf1: waveform)`` -> waveform - Output waveform.
+- ``offset(wf: waveform, amount: complex)`` -> waveform - Offset the input waveform by an amount.
+- ``scale(wf: waveform, factor: complex)`` -> waveform - Scale the input waveform by a factor.
+- ``conj(wf: waveform) -> waveform`` - Conjugate the input waveform.
+- ``re(wf: waveform) -> waveform`` - Real component of input waveform.
+- ``im(wf: waveform) -> waveform`` - Imaginary component of input waveform.
+- ``abs(wf: waveform) -> waveform`` - Transform waveform as norm of input. waveform
+- ``phase_shift(wf: waveform, ang: angle) -> waveform`` - Signal with relative phase, ang.
+
 Play instruction
 ----------------
 
-Pulses are scheduled using the ``play`` instruction. These instructions may
+Waveforms are scheduled using the ``play`` instruction. These instructions may
 only appear inside a ``defcal`` block!
 
 Play instructions have two required parameters:
 
-- a value of type ``pulse`` representing the pulse envelope
+- a value of type ``waveform`` representing the waveform envelope
 - the frame to use for the pulse
+- the channel on which to play the pulse
 
-.. code-block:: none
+.. code-block:: javascript
 
    // Play a 3 sample pulse on qubit 0's "drive" frame
-   play([1+0*j, 0+1*j, 1/sqrt(2)+1/sqrt(2)*j]) frameof("drive", %0);
+   play(tx0, [1+0*j, 0+1*j, 1/sqrt(2)+1/sqrt(2)*j], driveframe);
```

Comment:
Could we tie channels with frames in the form of virtual channels, eg., `vc chan = virtual(channel, frame)` from this point on a virtual channel is a frame+channel and all updates to the frame have a well-defined time.

Reply:
I like this. Also could remove the need to copy frames if we then tracked time via the virtual channels.